src/compiler/core.rs

```diff
@@ -1,13 +1,42 @@
-pub fn for_each<TItem, TReturn>(
-    array: &[TItem],
-    callback: &mut dyn FnMut(&TItem, usize) -> Option<TReturn>,
+use crate::SortedArray;
+
+pub fn for_each<
+    TCollection: IntoIterator,
+    TReturn,
+    TCallback: FnMut(TCollection::Item, usize) -> Option<TReturn>,
+>(
+    array: TCollection,
+    mut callback: TCallback,
 ) -> Option<TReturn> {
     array
-        .iter()
+        .into_iter()
         .enumerate()
         .find_map(|(index, item)| callback(item, index))
 }
 
+fn some<TItem>(array: &[TItem], predicate: Option<Box<dyn FnMut(&TItem) -> bool>>) -> bool {
```

Comment:
I think I should change things like this to be generic over the closure type rather than boxing the closure type - although here then if the caller is passing `None` for `predicate` the compiler would probably complain about not being able to resolve the closure generic?

Reply:
Yep, I've run into the same problem trying to make the inner value of an Option generic. I think I ended up just making a variants of the function rather than using an Option, but that's annoying.